config/reader.go

```diff
@@ -390,24 +390,36 @@ func processConfigLocalConfig(unprocessedConfig *Config, logger logging.Logger)
 	return processConfig(unprocessedConfig, false, logger)
 }
 
+// processConfig processes the config passed in. The config can be either JSON or gRPC derived.
+// If any part of this function errors, the function will exit and no part of the new config will be returned
+// until it is corrected.
 func processConfig(unprocessedConfig *Config, fromCloud bool, logger logging.Logger) (*Config, error) {
+	// Ensure validates the config but also substitutes in some defaults. Implicit dependencies for builtin resource
+	// models are not filled in until attributes are converted.
 	if err := unprocessedConfig.Ensure(fromCloud, logger); err != nil {
 		return nil, err
 	}
 
+	// we cached the unprocessed config, so make a copy before changing it too much. Also ensures validation
+	// happens again on resources, remotes, and modules.
 	cfg, err := unprocessedConfig.CopyOnlyPublicFields()
 	if err != nil {
 		return nil, errors.Wrap(err, "error copying config")
 	}
 
+	// Copy does not preserve ConfigFilePath since it preserves only JSON-exported fields and so we need
+	// to pass it along manually. ConfigFilePath needs to be preserved so the correct config watcher can
+	// be instantiated later in the flow.
+	cfg.ConfigFilePath = unprocessedConfig.ConfigFilePath
```

Comment:
why do we need to do this earlier?

Reply:
we don't need to, but figured it's a bit more straightforward to group the operations related to the copy together